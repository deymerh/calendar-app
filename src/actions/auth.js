import Swal from 'sweetalert2';
import { fetchWithoutToken, fetchWithToken } from '../helpers/fetch';
import { types } from '../types/types';

export const startLogin = (email, password) => {
  return async (dispatch) => {
    try {
      const res = await fetchWithoutToken('auth', { email, password }, 'POST');
      const body = await res.json();
      if (body.ok) {
        localStorage.setItem('token', body.token);
        localStorage.setItem('token-init-date', new Date().getTime());
        dispatch(login({
          uid: body.uid,
          name: body.name
        }));
      } else {
        Swal.fire('Error', body.msg, 'error')
      }
    } catch (error) {
      console.log(error);
    }
  }
}
export const startRegister = (name, email, password) => {
  return async (dispatch) => {
    try {
      const res = await fetchWithoutToken('auth/register', { name, email, password }, 'POST');
      const body = await res.json();
      if (body.ok) {
        localStorage.setItem('token', body.token);
        localStorage.setItem('token-init-date', new Date().getTime());
        dispatch(login({
          uid: body.uid,
          name: body.name
        }));
        Swal.fire('Ok', body.msg, 'success');
      } else {
        Swal.fire('Error', body.msg, 'error');
      }
    } catch (error) {
      console.log(error);
    }
  }
}
export const startChecking = () => {
  return async (dispatch) => {
    const res = await fetchWithToken('auth/renew');
    const body = await res.json();
    if (body.ok) {
      localStorage.setItem('token', body.token);
      localStorage.setItem('token-init-date', new Date().getTime());
      dispatch(login({
        uid: body.uid,
        name: body.name
      }));
    } else {
      dispatch(checkingSinish());
    }
  }
}

export const startLogout = () => {
  return (dispatch) => {
    localStorage.clear();
    dispatch(logout)
  }
}
const logout = () => ({ type: types.authLogout });

const login = (user) => ({
  type: types.authLogin,
  payload: user
});

const checkingSinish = () => ({ type: types.authCheckingFinish });