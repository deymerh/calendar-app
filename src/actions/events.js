import Swal from 'sweetalert2';
import { fetchWithToken } from '../helpers/fetch';
import { prepareEvents } from '../helpers/prepareEvents';
import { types } from '../types/types';

export const eventStartAddNew = (event) => {
  return async (dispatch, getState) => {
    const { name, uid } = getState().auth;
    try {
      const res = await fetchWithToken('events', event, 'POST');
      const body = await res.json();
      if (body.ok) {
        event.id = body.event.id;
        event.user = {
          _id: uid,
          name: name
        }
        dispatch(eventAddNew(event));
      }
    } catch (error) {
      console.log(error);
    }
  };
};

const eventAddNew = (event) => ({
  type: types.eventAddNew,
  payload: event
});

export const eventSetActive = (event) => ({
  type: types.eventSetActive,
  payload: event
});

export const eventClearActive = () => ({ type: types.eventClearActive });

export const eventStartDeleted = () => {
  return async (dispatch, getState) => {
    const { id } = getState().calendar.eventActive;
    try {
      const res = await fetchWithToken(`events/${id}`, {}, 'DELETE');
      const body = await res.json();
      if (body.ok) {
        dispatch(eventDeleted());
        Swal.fire('Ok', body.msg, 'success');
      } else {
        Swal.fire('Error', body.msg, 'error');
      }
    } catch (error) {
      console.log(error);
    }
  };
};

const eventDeleted = () => ({ type: types.eventDeleted });

export const startEventsLoaded = () => {
  return async (dispatch) => {
    const res = await fetchWithToken('events');
    const body = await res.json();
    const events = prepareEvents(body.events);
    dispatch(eventsLoaded(events));
  };
};

export const eventStartUpdate = (event) => {
  return async (dispath) => {
    try {
      const res = await fetchWithToken(`events/${event.id}`, event, 'PUT');
      const body = await res.json();
      console.log(body);
      if (body.ok) {
        dispath(eventUpdated(event));
      } else {
        Swal.fire('Error', body.msg, 'error');
      }
    } catch (error) {
      console.log(error);
    }
  }
}

const eventUpdated = (event) => ({
  type: types.eventUpdated,
  payload: event
});

const eventsLoaded = (events) => ({
  type: types.eventsLoaded,
  payload: events
});

export const eventLogout = () => ({ type: types.eventsLogout });