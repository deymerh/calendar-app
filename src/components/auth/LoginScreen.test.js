import '@testing-library/jest-dom';
import React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { render, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import moment from 'moment';
import { LoginScreen } from './LoginScreen';
import { startLogin, startRegister } from '../../actions/auth';
import Swal from 'sweetalert2';

jest.mock('../../actions/auth', () => ({
  startLogin: jest.fn(),
  startRegister: jest.fn(),
}));

jest.mock('sweetalert2', () => ({
  fire: jest.fn()
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initialState = {
  events: [
    {
      id: new Date().getTime(),
      title: 'Mi cumpleaños',
      start: moment().toDate(),
      end: moment().add(2, 'hours').toDate(),
      bgColor: '#fafafa',
      user: {
        _id: '12345',
        name: 'Deymer'
      }
    }
  ],
  eventActive: null
}

let store = mockStore(initialState);
store.dispatch = jest.fn();

describe('Test in the component [<LoginScreen />]', () => {

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('should do Snapshot the component [<LoginScreen />]', async () => {
    const { container } = render(
      <Provider store={store}>
        <LoginScreen />
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });

  test('should to do the login', async () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <LoginScreen />
      </Provider>
    );
    const btnLogin = await getByTestId('bntLogin');
    fireEvent.click(btnLogin);
    expect(startLogin).toHaveBeenCalledWith('deymerh@hotmail.com', '123456');
  });

  test('should do not the login and should to show the error message incorrect passwords', async () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <LoginScreen />
      </Provider>
    );
    const rPassword2 = await getByTestId('rPassword2');
    fireEvent.change(rPassword2, { target: { value: "" } });
    const btnRegister = await getByTestId('btnRegister');
    fireEvent.click(btnRegister);
    expect(startRegister).not.toHaveBeenCalled();
    expect(Swal.fire).toHaveBeenCalledWith('Error', 'Las contraseñas debe ser iguales', 'error');
  });

  test('should to do the register', async () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <LoginScreen />
      </Provider>
    );
    const rName = await getByTestId('rName');
    const rEmail = await getByTestId('rEmail');
    const rPassword1 = await getByTestId('rPassword1');
    const rPassword2 = await getByTestId('rPassword2');
    fireEvent.change(rName, { target: { value: 'Tiago' } });
    fireEvent.change(rEmail, { target: { value: 'tiago@tiago.com' } });
    fireEvent.change(rPassword1, { target: { value: '123456' } });
    fireEvent.change(rPassword2, { target: { value: '123456' } });
    const btnRegister = await getByTestId('btnRegister');
    fireEvent.click(btnRegister);
    expect(startRegister).toHaveBeenCalledWith('Tiago', 'tiago@tiago.com', '123456');
    expect(Swal.fire).not.toHaveBeenCalled();
  });

});
