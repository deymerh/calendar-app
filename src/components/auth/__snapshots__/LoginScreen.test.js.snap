// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in the component [<LoginScreen />] should do Snapshot the component [<LoginScreen />] 1`] = `
<div>
  <div
    class="container login-container"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6 login-form-1"
      >
        <h3>
          Ingreso
        </h3>
        <form>
          <div
            class="form-group"
          >
            <input
              class="form-control"
              name="lEmail"
              placeholder="Correo"
              type="text"
              value="deymerh@hotmail.com"
            />
          </div>
          <div
            class="form-group"
          >
            <input
              class="form-control"
              name="lPassword"
              placeholder="Contraseña"
              type="password"
              value="123456"
            />
          </div>
          <div
            class="form-group"
          >
            <input
              class="btnSubmit"
              data-testid="bntLogin"
              type="submit"
              value="Login"
            />
          </div>
        </form>
      </div>
      <div
        class="col-md-6 login-form-2"
      >
        <h3>
          Registro
        </h3>
        <form>
          <div
            class="form-group"
          >
            <input
              class="form-control"
              data-testid="rName"
              name="rName"
              placeholder="Nombre"
              type="text"
              value="Antonio"
            />
          </div>
          <div
            class="form-group"
          >
            <input
              class="form-control"
              data-testid="rEmail"
              name="rEmail"
              placeholder="Correo"
              type="email"
              value="deyer1227@gmail.com"
            />
          </div>
          <div
            class="form-group"
          >
            <input
              class="form-control"
              data-testid="rPassword1"
              name="rPassword1"
              placeholder="Contraseña"
              type="password"
              value="123456"
            />
          </div>
          <div
            class="form-group"
          >
            <input
              class="form-control"
              data-testid="rPassword2"
              name="rPassword2"
              placeholder="Repita la contraseña"
              type="password"
              value="123456"
            />
          </div>
          <div
            class="form-group"
          >
            <input
              class="btnSubmit"
              data-testid="btnRegister"
              type="submit"
              value="Crear cuenta"
            />
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
